### PR TITLE
Add benchmarks

### PR DESCRIPTION
**Pull request by clipperhouse**
Relating to this discussion: https://github.com/timbray/rfc9839/pull/12#discussion_r2310543442

I believed there was an allocation on `[]byte(s)`, but that appears incorrect, I not seeing an allocation. Glad to be wrong.

I created a “string native” version which is ~identical. No allocations in either case. The benchmarks are within a margin of error, _maybe_ the new one is 1% faster.

```
go test -bench=BenchmarkValidString -benchmem
goos: darwin
goarch: arm64
pkg: github.com/timbray/rfc9839
cpu: Apple M2
BenchmarkValidStringPrevious-8   	    1390	    767332 ns/op	 142.03 MB/s	       0 B/op	       0 allocs/op
BenchmarkValidString-8           	    1587	    756529 ns/op	 144.05 MB/s	       0 B/op	       0 allocs/op
PASS
ok  	github.com/timbray/rfc9839	2.643s
```

Will leave it to @timbray whether this implementation is interesting. I’m guessing not, as it’s more code for a possibly-illusory benefit. Feel free to keep the benchmarks but drop the impl.

The `testdata` is from my [other Unicode package](https://github.com/clipperhouse/uax29), it is basically lorem ipsum for a sampling of languages.

### CODE REVIEW

**Comment by clipperhouse** [comment]
Updated with a more accurate bytes count, since we are doing 3 tests per bench iteration. I believe the following MB/s are more representative:

```
go test -bench=BenchmarkValidString -benchmem
goos: darwin
goarch: arm64
pkg: github.com/timbray/rfc9839
cpu: Apple M2
BenchmarkValidStringPrevious-8   	    1390	    767189 ns/op	 426.16 MB/s	       0 B/op	       0 allocs/op
BenchmarkValidString-8           	    1598	    753874 ns/op	 433.68 MB/s	       0 B/op	       0 allocs/op
PASS
ok  	github.com/timbray/rfc9839	2.650s
```

**Comment by timbray** [comment]
Your implementation of `isStringInSubset` is identical to mine, with similar results.  I think I'll stand pat on the code for now, shorter is better.  Tell me a little about your test files?  I also have a decent but all-ASCII benchmark which BTW produced the same finding of no-significant-difference.  It'd be useful to have a test file with a lot of Chinese text in it, which would stress the simple-but-fast algorithm a little harder.

**Comment by clipperhouse** [comment]
Cool. The test files (only using `sample.txt` in this case) were me manually copying some English text from Wikipedia and then IIRC using an online lorem ipsum generator that supported multiple languages. I believe you’ll find Chinese in there.

Tempted to ask an LLM to generate something better. :)

**Comment by timbray** [comment]
I like your benchmark better than mine, if you'd back out the change to unichars.go I'll merge.

**Comment by clipperhouse** [comment]
Cool, done.

File: unichars_bench_test.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	bytes := len(file)
+	b.SetBytes(int64(bytes))
+	b.ReportAllocs()
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		Assignables.ValidUtf8(file)
```
Comment (timbray) [inline unichars_bench_test.go:21]: If I were writing this, I'd actually check the return code from the Valid* calls.  Just because I'm paranoid and this big chunk of text might expose a corner case that the white-box unit tests don't.  Any reason not to?

**Review by timbray — COMMENTED** [review]
see inline

File: unichars_bench_test.go (hunk 1 above)
Comment (clipperhouse) [reply unichars_bench_test.go:21]: Good idea. For my tastes, I prefer to separate the concerns of testing and benchmarking.

In the latest commit, I created a better sample file, and added tests. Also added tests using the invalid UTF-8 file. Left the benchmarks as-is. Have a look.

**Review by timbray — APPROVED** [review]
